src/main/kotlin/org/rust/ide/docs/RustDocumentationProvider.kt

```diff
@@ -60,15 +60,35 @@ class RustDocumentationProvider : AbstractDocumentationProvider() {
         }
 
         val signatureStart = signatureStartElement?.startOffsetInParent ?: 0
-        val signatureEnd = block?.startOffsetInParent ?: textLength
+
+        // pick last of: where clause, return type, or closing brace of the parameters
+        // if all else fails, drop down to the length of the current element
+        val signatureEnd = if (block != null)
+            block!!.startOffsetInParent
+        else if (whereClause != null)
+            whereClause!!.let { it.startOffsetInParent + it.textLength }
+        else if (retType != null)
+            retType!!.let { it.startOffsetInParent + it.textLength }
+        else if (parameters != null)
+            parameters!!.let { it.startOffsetInParent + it.textLength }
+        else
+            textLength
 
         val identStart = identifier.startOffsetInParent
         val identEnd = identStart + identifier.textLength
         check(signatureStart <= identStart && identStart <= identEnd &&
             identEnd <= signatureEnd && signatureEnd <= textLength)
 
         val beforeIdent = text.substring(signatureStart, identStart).escaped
-        val afterIdent = text.substring(identEnd, signatureEnd).trimEnd().escaped
+        val afterIdent = text.substring(identEnd, signatureEnd)
```

Comment:
@matklad This new change handles the case where the params, return types and where clause are all spread out. But, it's pretty ugly.

I think the right thing to do is normalize every signature element and then concatenate them with spaces. Unfortunately I'm not sure how best to do this.